### DISCUSSION TITLE
The result of a steady free thermal expansion is so weird...

### DISCUSSION THREAD

**Discussion by Wang-Yihu** [idaholab/moose#28205] [Q&A Modules: Solid mechanics] [answered]
Let's consider a beam, with a d = 0.2 side length square cross section, and l = 10.0 axial length. Thermal expansion coefficient is α = 2.0e-5. Stress free temperature and reference temperature is T_r = 300. Temperature is T = 500. All boundaries are free. There are no loads except thermal expansion. It's a very easy question...
![image](https://github.com/user-attachments/assets/694583fd-f4fa-4f0f-9e9f-8771474ace38)

So we have this input file
```
d = 200.0e-3
l = 10.0
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 10
    ny = 10
    xmin = ${fparse -d/2}
    xmax = ${fparse d/2}
    ymin = ${fparse -d/2}
    ymax = ${fparse d/2}
  []
  
  [extrude1]
    type = AdvancedExtruderGenerator
    input = gmg
    heights = ${fparse l}
    num_layers = 100
    direction = '0 0 1'
    bottom_boundary = '99'
    top_boundary = '66'
  []
[]

[Materials]
  [./elasticity_tensor_steel]
    youngs_modulus = 210e9
    poissons_ratio = 0.28
    type = ComputeIsotropicElasticityTensor
  [../]
  [expansion_steel]
    type = ComputeMeanThermalExpansionFunctionEigenstrain
    thermal_expansion_function = 2e-5
    thermal_expansion_function_reference_temperature = 300.0
    stress_free_temperature = 300.0
    temperature = temp_jiegou
    eigenstrain_name = thermal
  []
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]

[AuxVariables]
  [temp_jiegou]
    family = LAGRANGE
    order = FIRST
  []
[]

[AuxKernels]
  [changshu_wendu]
    type = ConstantAux
    variable = temp_jiegou
    value = 500.0
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    add_variables = true
    eigenstrain_names = thermal
    generate_output = 'vonmises_stress stress_zz'
    strain = FINITE
    temperature = temp_jiegou
  []
[]

[BCs]
[]


[Executioner]
  type = Steady
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'
  nl_max_its = 50000
[]

[Postprocessors]
  [max_x]
    type = NodalMaxValue
    variable = disp_x
  []
  [max_y]
    type = NodalMaxValue
    variable = disp_y
  []
  [max_z]
    type = NodalMaxValue
    variable = disp_z
  []
[]

[Outputs]
  exodus = true
[]
```
However, the result is...
![image](https://github.com/user-attachments/assets/0ee9cee7-9cbc-4fd5-a999-dadc1abf4b53)

The result is so weird, right? The displacement from MOOSE is so huge...In my problem, ΔT = T - T_r = 200. Thermal expansion strain ε_th = αΔT = 0.004. It seems that max disp is only ε_th*l = 0.04...The magnitude disp result from MOOSE is e+01, e+02...
Maybe in MOOSE steady solid mechanics calculation, there must be at least a BC to prevent the rigid body movement...So at least I need to set a disp = 0 in the centroid node of the model??
Looking forward to your reply...

**Comment by GiudGiud** [comment, marked as answer]
Hello

There is nothing holding the beam. So it is just moving around, in the nullspace of all the allowed solutions (moving the beam by 1m to the left does not make the residual increase at all!)

You were on the right track with preventing rigid body movement.
you have to add a bunch of dirichlet BCs (in x y z) to kill translational and rotational modes which you do not care about

Guillaume
